executor/trace.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/opentracing/basictracer-go"
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/pingcap/tidb/plan"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/tracing"
+	"golang.org/x/net/context"
+)
+
+// TraceExec represents a root executor of trace query.
+type TraceExec struct {
+	baseExecutor
+	// CollectedSpans collects all span during execution. Span is appended via
+	// callback method which passes into tracer implementation.
+	CollectedSpans []basictracer.RawSpan
+	// exhausted being true means there is no more result.
+	exhausted bool
+	// plan is the real query plan and it is used for building real query's executor.
+	plan plan.Plan
+	// rootTrace represents root span which is father of all other span.
+	rootTrace opentracing.Span
+
+	childrenResults []*chunk.Chunk
+}
+
+// buildTrace builds a TraceExec for future executing. This method will be called
+// at build().
+func (b *executorBuilder) buildTrace(v *plan.Trace) Executor {
+	e := &TraceExec{
+		baseExecutor: newBaseExecutor(b.ctx, v.Schema(), v.ExplainID()),
+	}
+
+	pp, _ := v.StmtPlan.(plan.PhysicalPlan)
+	e.children = make([]Executor, 0, len(pp.Children()))
+	for _, child := range pp.Children() {
+		switch p := child.(type) {
+		case *plan.PhysicalTableReader, *plan.PhysicalIndexReader, *plan.PhysicalIndexLookUpReader, *plan.PhysicalHashAgg, *plan.PhysicalProjection, *plan.PhysicalStreamAgg, *plan.PhysicalSort:
+			e.children = append(e.children, b.build(p))
+		default:
+			b.err = errors.Errorf("%v is not supported", child)
+		}
+
+	}
+
+	return e
+}
+
+// Open opens a trace executor and it will create a root trace span which will be
+// used for the following span in a relationship of `ChildOf` or `FollowFrom`.
+// for more details, you could refer to http://opentracing.io
+func (e *TraceExec) Open(ctx context.Context) error {
```

Comment:
seems that we can not trace the **building** and **optimization** time of a query, only the **execution** time can be traced?

Reply:
It is only for now. I will file another PR to support this.

Reply:
Since the `TraceExec` has no child, I think we can just leave this function to be empty.